### ISSUE TITLE
Handling Unsigned, Hex-bit, Hexdip, Sixbit fields possible in python script.?

### ISSUE THREAD

**Issue by AnveshAeturi** [aws-samples/mainframe-data-utilities#9]
Hi Luis, we are working with the mainframe data utilities repo for a data conversion in our files. While it is working for most of the common cobol fields, we have a few fields with data-type as **"Packed unsigned", "binary unsigned", "Hex-bit", "Hexdisp" & "Six-bit"**. As we understand, these fields are coming from Assembler in Mainframe and represented using an **Easytrieve layout**, for which there are no equivalent data types in Cobol for it to be put on a cobol layout.

Is there a way we can handle these datatypes or is an enhancement possible on the Python scripts to handle them or is it handling any of the mentioned data types?

**Comment by luisdant** [comment]
Hi Anvesh,

Can you please share the Easytrieve representation of these fields?

thx

**Comment by AnveshAeturi** [comment]
[Easyterieve_Layout_sample.xlsx](https://github.com/aws-samples/mainframe-data-utilities/files/9597068/Easyterieve_Layout_sample.xlsx)


This is  the Excel which we created from the Easytrieve layout. only sample fields are added here.

**Comment by luisdant** [comment]
Hi Anvesh.

The only unsupported field in the provided layout is the 1 byte binary, which doesn't exist in Cobol.

For now I'm adding this as a limitation in the read me file.

As a workaround you can define your field as "PIC X", parse the copybook into json using the copybook parser, and change the field definition from "ch" to "bi".

Regards

Luis

**Comment by AnveshAeturi** [comment]
Hi Luis, Thanks For your inputs and Help you have done. Really appreciate it. 

As we seen and looked into Data stored in Mainframe's we found a lot of Data Types apart from which were asked in this Ticket. we followed the approach  you have done ,like splitting the file into lines (length of lines) and lines into chunks of fields based on the bytes and offset, and then converting the chunks to ASCII Data based on the types. Just to avoid the Redefines problem we parsed the copybook manually into csv file as it is one time effort.

After understanding the Data Stored in the mainframes, our understand as follows for the above asked fields :
1.  **Unsigned Decimals**:  we still have to use bytes.hex() + decimal (integer value ) extra parameter will be there.  
> EX: after hex conversion if we got the 22354 as a value and decimal parameter as 2, then final output will be 223.54 

2. **Hex Bit**: we need to convert the bytes to hex first and based on the scale and num_of _bits the transform as follows:
>EX: where Scale = 16 and num_of_bits = 8
> bin(int(bytes.hex(), scale))[2:].zfill(num_of_bits)

3. **Hex dip:**  even though it sounds different its still the same just have to convert to hex
> Ex: bytes.hex()

4. **Six bit:** Six bit something which is combination /transformations of some steps which involves more than 5 types of conversion which is totally unique.
